keyboards/gmmk/pro/ansi/keymaps/andrebrait/keymap.c

```diff
@@ -0,0 +1,171 @@
+        _______, _______, _______,                            _______,                            _______, _______, _______, RGB_SPD, RGB_RMOD, RGB_SPI
+    ),
+
+
+};
+
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (clockwise) {
+      tap_code(KC_VOLU);
+    } else {
+      tap_code(KC_VOLD);
+    }
+    return true;
+}
+
+#ifdef RGB_MATRIX_ENABLE
+/* Flag to indicate that the RGB Matrix is only ON because we need it for CAPS LOCK */
+bool rgb_matrix_turned_on_by_caps = false;
+
+bool led_update_user(led_t led_state) {
+    if (led_state.caps_lock) {
+        if (!rgb_matrix_is_enabled()) {
+            /* Turn ON the RGB Matrix for CAPS LOCK */
+            rgb_matrix_turned_on_by_caps = true;
+            rgb_matrix_enable_noeeprom();
+        }
+    } else if (rgb_matrix_turned_on_by_caps) {
+        /* RGB Matrix was only ON because of CAPS LOCK. Turn it OFF. */
+        rgb_matrix_turned_on_by_caps = false;
+        rgb_matrix_disable_noeeprom();
+    }
+    return true;
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    switch (keycode) {
+        case RGB_MOD:
+        case RGB_RMOD:
+        case RGB_HUI:
+        case RGB_HUD:
+        case RGB_SAI:
+        case RGB_SAD:
+        case RGB_VAI:
+        case RGB_VAD:
+        case RGB_SPI:
+        case RGB_SPD:
+            if (record->event.pressed) {
+                if (rgb_matrix_turned_on_by_caps) {
+                    /*
+                    Prevent changes to the RGB Matrix settings while it's supposed to be OFF.
+                    Returning true will also persist the current state to EEPROM and we don't want
+                    that when the RGB Matrix is only ON because of CAPS LOCK.
+                    */
+                    return false;
+                }
+            }
+            break;
+        case RGB_TOG:
+            if (record->event.pressed) {
+                if (rgb_matrix_is_enabled()) {
+                    if (rgb_matrix_turned_on_by_caps) {
+                        /*
+                        Transitioning (fake) OFF -> (real) ON.
+                        Persist current state to EEPROM (it's already ON)
+                        */
+                        rgb_matrix_turned_on_by_caps = false;
+                        eeconfig_update_rgb_matrix();
+                        return false;
+                    } else if (host_keyboard_led_state().caps_lock) {
+                        /*
+                        Transitioning (real) ON -> (fake) OFF.
+                        Persist OFF to EEPROM but keep it ON
+                        */
+                        rgb_matrix_turned_on_by_caps = true;
+                        rgb_matrix_disable_noeeprom();
+                        eeconfig_update_rgb_matrix();
+                        rgb_matrix_enable_noeeprom();
+                        return false;
+                    }
+                }
+            }
+            break;
```

Comment:
```suggestion
        case RGB_TOG:
            if (record->event.pressed) {
                switch (rgb_matrix_get_flags()) {
                    case LED_FLAG_ALL: {
                        rgb_matrix_set_flags(LED_FLAG_NONE);
                        // keyboard_config.rgb_matrix_enable = false;
                        rgb_matrix_set_color_all(0, 0, 0);
                    } break;
                    default: {
                        rgb_matrix_set_flags(LED_FLAG_ALL);
                        // keyboard_config.rgb_matrix_enable = true;
                    } break;
                }
                //eeconfig_update_kb(keyboard_config.raw);
            }
            return false;
```

Reply:
So, this would allow me to turn on the LEDs I want without touching anything in the RGB Matrix struct or depending on the matrix being active?

Reply:
Unfortunately, this doesn't work.

As soon as I disable the matrix, whatever state the CAPS LOCK LEDs were in will stay that way until the matrix is enabled again.
Returning `false` on `process_record_user` causes the state not to be persisted to the EEPROM, so the keyboard will not retain the state of the matrix upon restart.
Interestingly, calling `rgb_matrix_disable()` and `rgb_matrix_enable()` also doesn't cause the state to be persisted, which is why I was resorting to enabling/disabling the matrix without persisting and then doing the persistence manually.

I was trying to get this to work since last week but this was really the only way I could find. Everything else had some issue.

Do you have any other idea how I can achieve this without doing what I did?

The goals are:
1. The CAPS LOCK LEDs must be able to come up at any moment, with or without the RGB Matrix being active
2. The state of CAPS LOCK must be persisted across restarts
3. The state of the RGB Matrix must be persisted across restarts
4. While the RGB Matrix is disabled, no key combinations should be able to change its state
    - Specially in a way that persists to the EEPROM
5. It should be possible to persist (and get the keyboard to restart with exactly the same settings) for any combination of CAPS LOCK LEDs state and RGB Matrix state

Reply:
Yeah, the LEDs don't clean up automatically when you disable them this way, so you have to manually clean them up.